File: script-files/measures/outputsContinuous/gamma_3p.js
jStat = require("../node_modules/jstat");

dists = {
    gamma_3p: {
        measurements: {
            nonCentralMoments: function (k, alpha, beta, loc) {
                return beta ** k * (jStat.gammafn(k + alpha) / (jStat.factorial(alpha - 1)))
            },
            centralMoments: function (k, alpha, beta, loc) {
                const miu1 = this.nonCentralMoments(1, alpha, beta, loc);
                const miu2 = this.nonCentralMoments(2, alpha, beta, loc);
                const miu3 = this.nonCentralMoments(3, alpha, beta, loc);
                const miu4 = this.nonCentralMoments(4, alpha, beta, loc);

                let result;
                switch (k) {
                    case 1: result = 0; break;
                    case 2: result = miu2 - miu1 ** 2; break;
                    case 3: result = miu3 - 3 * miu1 * miu2 + 2 * miu1 ** 3; break;
                    case 4: result = miu4 - 4 * miu1 * miu3 + 6 * (miu1 ** 2) * miu2 - 3 * (miu1 ** 4); break;
                };
                return result
            },
            stats: {
                mean: function (alpha, beta, loc) {
                    const miu1 = dists.gamma_3p.measurements.nonCentralMoments(1, alpha, beta, loc);
                    return loc + miu1;
                },
                variance: function (alpha, beta, loc) {
                    const miu1 = dists.gamma_3p.measurements.nonCentralMoments(1, alpha, beta, loc);
                    const miu2 = dists.gamma_3p.measurements.nonCentralMoments(2, alpha, beta, loc);
                    return miu2 - miu1 ** 2;
                },
                standardDeviation: function (alpha, beta, loc) {
                    return Math.sqrt(this.variance(alpha, beta, loc));
                },
                skewness: function (alpha, beta, loc) {
                    const central_miu3 = dists.gamma_3p.measurements.centralMoments(3, alpha, beta, loc);
                    return central_miu3 / (this.standardDeviation(alpha, beta, loc) ** 3);
                },
                kurtosis: function (alpha, beta, loc) {
                    const central_miu4 = dists.gamma_3p.measurements.centralMoments(4, alpha, beta, loc);
                    return central_miu4 / (this.standardDeviation(alpha, beta, loc) ** 4);
                },
                median: function (alpha, beta, loc) {
                    return dists.gamma_3p.measurements.ppf(0.5, alpha, beta, loc);
                },
                mode: function (alpha, beta, loc) {
                    return beta * (alpha - 1) + loc;
                },
            },
        }
    }
}
console.log(dists.gamma_3p.measurements.stats.mean(25, 2, 100))
console.log(dists.gamma_3p.measurements.stats.variance(25, 2, 100))
console.log(dists.gamma_3p.measurements.stats.standardDeviation(25, 2, 100))
console.log(dists.gamma_3p.measurements.stats.skewness(25, 2, 100))
console.log(dists.gamma_3p.measurements.stats.kurtosis(25, 2, 100))
// console.log(dists.gamma_3p.measurements.stats.median(25, 2, 100))
console.log(dists.gamma_3p.measurements.stats.mode(25, 2, 100))

// mean_value: 150
// variance_value: 100.00000000000136
// standard_deviation_value: 10.000000000000068
// skewness_value: 0.3999999999998755
// kurtosis_value: 3.2400000000010305
// median_value: 149.33493673397683
// mode_value: 148

File: continuous/distributions/gamma_3p.py
import math
import scipy.special as sc
import scipy.stats


class GAMMA_3P:
    """
    Gamma distribution
    https://en.wikipedia.org/wiki/Gamma_distribution
    Compendium of Common Probability Distributions (pag.39) ... Michael P. McLaughlin
    """

    def __init__(self, measurements):
        self.parameters = self.get_parameters(measurements)
        self.alpha = self.parameters["alpha"]
        self.beta = self.parameters["beta"]
        self.loc = self.parameters["loc"]

    def cdf(self, x: float) -> float:
        """
        Cumulative distribution function
        Calculated using the definition of the function
        Alternative: quadrature integration method
        """
        ## Method 1: Integrate PDF function
        # result, error = scipy.integrate.quad(self.pdf, 0, x)
        # print(result)

        ## Method 2: Scipy Gamma Distribution class
        # result = scipy.stats.gamma.cdf(x, a=self.alpha, scale=self.beta)
        # print(result)
        result = sc.gammainc(self.alpha, (x - self.loc) / self.beta)
        return result

    def pdf(self, x: float) -> float:
        """
        Probability density function
        Calculated using definition of the function in the documentation
        """
        # result = ((self.beta**-self.alpha) * ((x - self.loc) ** (self.alpha - 1)) * math.exp(-((x - self.loc) / self.beta))) / math.gamma(self.alpha)
        result = scipy.stats.gamma.pdf(x, self.alpha, loc=self.loc, scale=self.beta)
        return result

    def get_num_parameters(self) -> int:
        """
        Number of parameters of the distribution
        """
        return len(self.parameters)

    def parameter_restrictions(self) -> bool:
        """
        Check parameters restrictions
        """
        v1 = self.alpha > 0
        v2 = self.beta > 0
        return v1 and v2

    def get_parameters(self, measurements) -> dict[str, float | int]:
        """
        Calculate proper parameters of the distribution from sample measurements.
        The parameters are calculated by formula.

        Parameters
        ==========
        measurements: MEASUREMESTS
            attributes: mean, std, variance, skewness, kurtosis, median, mode, min, max, length, num_bins, data

        Returns
        =======
        parameters : dict
            {"alpha":  * , "beta":  * }
        """
        alpha = (2 / measurements.skewness) ** 2
        beta = math.sqrt(measurements.variance / alpha)
        loc = measurements.mean - alpha * beta
        parameters = {"alpha": alpha, "loc": loc, "beta": beta}
        return parameters


if __name__ == "__main__":
    ## Import function to get measurements
    import sys

    sys.path.append("../measurements")
    from measurements_continuous import MEASUREMENTS_CONTINUOUS

    ## Import function to get measurements
    def get_data(direction):
        sample_distribution_file = open(direction, "r")
        data = [float(x.replace(",", ".")) for x in sample_distribution_file.read().splitlines()]
        return data

    ## Distribution class
    path = "../data/data_gamma_3p.txt"
    data = get_data(path)
    measurements = MEASUREMENTS_CONTINUOUS(data)
    distribution = GAMMA_3P(measurements)

    print(distribution.get_parameters(measurements))
    print(distribution.cdf(measurements.mean))
    print(distribution.pdf(measurements.mean))
